File: sdk/typescript/src/embedding_vector_store_conformance.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";
import { afterEach, describe, expect, it, vi } from "vitest";
import { EmbeddingClient } from "./embedding";
import { VectorStoreClient } from "./vector_store";

const realFetch = globalThis.fetch;
// fileURLToPath(import.meta.url) 取目录：bun + node/vitest 双 runtime 通用
// （不用 bun 专属 import.meta.dir——在 node/vitest 下为 undefined 会抛错）。
const HERE = dirname(fileURLToPath(import.meta.url));
const SHARED_FIXTURES = resolve(HERE, "..", "..", "shared-fixtures");

function loadFixture(name: string): any {
  return JSON.parse(readFileSync(resolve(SHARED_FIXTURES, name), "utf-8"));
}

function setEnv() {
  process.env.KS_GATEWAY_URL = "http://gw";
  process.env.KS_RELAY_TOKEN = "tk";
  process.env.KS_EMBEDDING_MODEL = "bge-m3";
}

afterEach(() => {
  globalThis.fetch = realFetch;
  delete process.env.KS_GATEWAY_URL;
  delete process.env.KS_RELAY_TOKEN;
  delete process.env.KS_EMBEDDING_MODEL;
});

describe("Embedding / VectorStore conformance fixtures", () => {
  it("EmbeddingClient matches embeddings_v1.json", async () => {
    const fixture = loadFixture("embeddings_v1.json");
    const captured: any = {};
    setEnv();
    globalThis.fetch = vi.fn(async (_url: any, init: any) => {
      captured.method = init.method;
      captured.path = "/v1/mcp/relay/embeddings";
      captured.body = JSON.parse(init.body);
      return new Response(JSON.stringify(fixture.response), { status: 200, headers: { "Content-Type": "application/json" } });
    }) as any;

    const got = await new EmbeddingClient().embed("hello");

    expect(captured).toEqual(fixture.request);
    expect(got.dense).toEqual([0.1, 0.2]);
    expect(got.tokens).toBe(2);
    expect(got.sparse).toEqual({ "100": 0.5, "7": 0.25 });
  });

  it("VectorStoreClient matches vector_store_search_text.json", async () => {
    const fixture = loadFixture("vector_store_search_text.json");
    const captured: any = {};
    setEnv();
    globalThis.fetch = vi.fn(async (_url: any, init: any) => {
      captured.method = init.method;
      captured.path = "/v1/mcp/relay/vector_store/search_text";
      captured.body = JSON.parse(init.body);
      return new Response(JSON.stringify(fixture.response), { status: 200, headers: { "Content-Type": "application/json" } });
    }) as any;

    const got = await new VectorStoreClient(new EmbeddingClient(), "documents").searchText("hello", { topK: 5 });

    expect(captured).toEqual(fixture.request);
    expect(got[0]?.id).toBe("doc1");
    expect(got[0]?.payload).toEqual({ doc_id: "doc1" });
  });
});

File: sdk/go/ksapp/embedding_vector_store_conformance_test.go
package ksapp

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

type relayFixture struct {
	Request struct {
		Method string         `json:"method"`
		Path   string         `json:"path"`
		Body   map[string]any `json:"body"`
	} `json:"request"`
	Response map[string]any `json:"response"`
}

func loadRelayFixture(t *testing.T, name string) relayFixture {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "..", "shared-fixtures", name))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var f relayFixture
	if err := json.Unmarshal(raw, &f); err != nil {
		t.Fatalf("decode fixture: %v", err)
	}
	return f
}

func TestEmbeddingClientConformanceFixture(t *testing.T) {
	f := loadRelayFixture(t, "embeddings_v1.json")
	var captured map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != f.Request.Method || r.URL.Path != f.Request.Path {
			t.Fatalf("request %s %s", r.Method, r.URL.Path)
		}
		if err := json.NewDecoder(r.Body).Decode(&captured); err != nil {
			t.Fatal(err)
		}
		_ = json.NewEncoder(w).Encode(f.Response)
	}))
	defer srv.Close()

	c := &EmbeddingClient{gatewayURL: srv.URL, relayToken: "tk", model: "bge-m3", httpClient: srv.Client()}
	got, err := c.Embed(context.Background(), "hello")
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(captured, f.Request.Body) {
		t.Fatalf("body=%#v want %#v", captured, f.Request.Body)
	}
	if got.Tokens != 2 || !reflect.DeepEqual(got.Dense, []float32{0.1, 0.2}) {
		t.Fatalf("unexpected embedding result: %#v", got)
	}
	if !reflect.DeepEqual(got.Sparse, map[string]float32{"100": 0.5, "7": 0.25}) {
		t.Fatalf("unexpected sparse embedding: %#v", got.Sparse)
	}
}

func TestVectorStoreClientConformanceFixture(t *testing.T) {
	f := loadRelayFixture(t, "vector_store_search_text.json")
	var captured map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != f.Request.Method || r.URL.Path != f.Request.Path {
			t.Fatalf("request %s %s", r.Method, r.URL.Path)
		}
		if err := json.NewDecoder(r.Body).Decode(&captured); err != nil {
			t.Fatal(err)
		}
		_ = json.NewEncoder(w).Encode(f.Response)
	}))
	defer srv.Close()

	emb := &EmbeddingClient{gatewayURL: srv.URL, relayToken: "tk", model: "bge-m3", httpClient: srv.Client()}
	got, err := newVectorStoreClient(emb, "documents").SearchText(context.Background(), "hello", SearchOptions{TopK: 5})
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(captured, f.Request.Body) {
		t.Fatalf("body=%#v want %#v", captured, f.Request.Body)
	}
	if len(got) != 1 || got[0].ID != "doc1" || got[0].Payload["doc_id"] != "doc1" {
		t.Fatalf("unexpected search result: %#v", got)
	}
}

func TestLLMChatIntentConformanceFixture(t *testing.T) {
	f := loadRelayFixture(t, "relay_chat_intent.json")
	var captured map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != f.Request.Method || r.URL.Path != f.Request.Path {
			t.Fatalf("request %s %s", r.Method, r.URL.Path)
		}
		if err := json.NewDecoder(r.Body).Decode(&captured); err != nil {
			t.Fatal(err)
		}
		_ = json.NewEncoder(w).Encode(f.Response)
	}))
	defer srv.Close()

	c := &LLMClient{gatewayURL: srv.URL, relayToken: "tk", httpClient: srv.Client()}
	if _, err := c.Chat(context.Background(), ChatRequest{
		Messages:            []Message{{Role: "user", Content: "hi"}},
		Tier:                "flagship",
		RequireCapabilities: []string{"vision"},
		Reasoning:           "on",
	}); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(captured, f.Request.Body) {
		t.Fatalf("intent wire body 不一致:\n got=%#v\nwant=%#v", captured, f.Request.Body)
	}
}
